src/gha_logs.rs

```diff
@@ -99,34 +102,64 @@ async fn process_logs(
 
     let log_uuid = format!("{owner}/{repo}${log_id}");
 
-    let logs = 'logs: {
+    let ((job, tree_roots), logs) = &*'logs: {
         if let Some(logs) = ctx.gha_logs.write().await.get(&log_uuid) {
             tracing::info!("gha_logs: cache hit for {log_uuid}");
             break 'logs logs;
         }
 
         tracing::info!("gha_logs: cache miss for {log_uuid}");
-        let logs = ctx
-            .github
-            .raw_job_logs(
-                &github::IssueRepository {
-                    organization: owner.to_string(),
-                    repository: repo.to_string(),
-                },
-                log_id,
-            )
-            .await
-            .context("unable to get the raw logs")?;
 
-        let json_logs = serde_json::to_string(&*logs).context("unable to JSON-ify the raw logs")?;
+        let repo = github::IssueRepository {
+            organization: owner.to_string(),
+            repository: repo.to_string(),
+        };
+
+        let job_and_tree_roots = async {
+            let job = ctx
+                .github
+                .workflow_run_job(&repo, log_id)
+                .await
+                .context("unable to fetch job details")?;
+            let trees = ctx
+                .github
+                .repo_git_trees(&repo, &job.head_sha)
+                .await
+                .context("unable to fetch git tree for the repository")?;
+
+            let tree_roots = trees
+                .tree
+                .iter()
+                .filter_map(|t| (t.object_type == "tree").then_some(&t.path))
+                .join("|");
+
+            anyhow::Result::<_>::Ok((job, tree_roots))
+        };
+
+        let logs = async {
+            let logs = ctx
+                .github
+                .raw_job_logs(&repo, log_id)
+                .await
+                .context("unable to get the raw logs")?;
+
+            let json_logs =
+                serde_json::to_string(&*logs).context("unable to JSON-ify the raw logs")?;
+
+            anyhow::Result::<_>::Ok(json_logs)
+        };
+
+        let (job_and_tree_roots, logs) = futures::join!(job_and_tree_roots, logs);
+        let job_and_tree_roots_and_jobs = (job_and_tree_roots?, logs?);
 
         ctx.gha_logs
             .write()
             .await
-            .put(log_uuid.clone(), json_logs.into())
+            .put(log_uuid.clone(), job_and_tree_roots_and_jobs.into())
```

Comment:
```suggestion
            .put(log_uuid.clone(), job_and_tree_roots_and_logs.into())
```
? But we should really create some named struct with fields here.. e.g. `JobData`. And `ctx.gha_logs` could also be renamed e.g. to `gha_jobs`.